src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1012 @@
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	int			 rc = 0;
+
+	rc = agg_prep_sgl(entry);
+	if (rc)
+		goto out;
+	recx.rx_idx = entry->ae_cur_stripe.as_stripenum * k * len;
+	recx.rx_nr = k * len;
+	iod.iod_name = entry->ae_akey;
+	iod.iod_type = DAOS_IOD_ARRAY;
+	iod.iod_size = entry->ae_rsize;
+	iod.iod_nr = 1;
+	iod.iod_recxs = &recx;
+	entry->ae_sgl.sg_nr = 1;
+	entry->ae_sgl.sg_iovs[AGG_IOV_DATA].iov_len =
+						len * k * entry->ae_rsize;
+
+	agg_param = container_of(entry, struct ec_agg_param, ap_agg_entry);
+	rc = vos_obj_fetch(agg_param->ap_cont_handle, entry->ae_oid,
+			   entry->ae_cur_stripe.as_hi_epoch,
+			   VOS_OF_FETCH_RECX_LIST, &entry->ae_dkey, 1, &iod,
+			   &entry->ae_sgl);
+	if (rc)
+		D_ERROR("vos_obj_fetch failed: "DF_RC"\n", DP_RC(rc));
+	entry->ae_sgl.sg_nr = AGG_IOV_CNT;
+out:
+	return rc;
+}
+
+/* Xstream offload function for encoding new parity from full stripe of
+ * replicas.
+ */
+static void
+agg_encode_full_stripe_ult(void *arg)
+{
+	struct ec_agg_stripe_ud	*stripe_ud =
+					(struct ec_agg_stripe_ud *)arg;
+	struct ec_agg_entry	*entry = stripe_ud->asu_agg_entry;
+	unsigned int		 len = entry->ae_oca->u.ec.e_len;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	unsigned int		 p = entry->ae_oca->u.ec.e_p;
+	unsigned int		 cell_bytes = len * entry->ae_rsize;
+	unsigned char		*data[k];
+	unsigned char		*parity_bufs[p];
+	unsigned char		*buf;
+	int			 i, rc = 0;
+
+	buf = entry->ae_sgl.sg_iovs[AGG_IOV_DATA].iov_buf;
+	for (i = 0; i < k; i++)
+		data[i] = &buf[i * cell_bytes];
+
+	buf = entry->ae_sgl.sg_iovs[AGG_IOV_PARITY].iov_buf;
+	for (i = p - 1; i >= 0; i--)
+		parity_bufs[i] = &buf[i * cell_bytes];
+
+	if (entry->ae_codec == NULL)
+		entry->ae_codec =
+		obj_ec_codec_get(daos_obj_id2class(entry->ae_oid.id_pub));
+	ec_encode_data(cell_bytes, k, p, entry->ae_codec->ec_gftbls, data,
+		       parity_bufs);
+
+	ABT_eventual_set(stripe_ud->asu_eventual, (void *)&rc, sizeof(rc));
+}
+
+/* Encodes a full stripe. Called when replicas form a full stripe.
+ */
+static int
+agg_encode_full_stripe(struct ec_agg_entry *entry)
+{
+	struct ec_agg_stripe_ud		stripe_ud = { 0 };
+	int				*status;
+	int				rc = 0;
+
+	stripe_ud.asu_agg_entry = entry;
+	rc = ABT_eventual_create(sizeof(*status), &stripe_ud.asu_eventual);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+	rc = dss_ult_create(agg_encode_full_stripe_ult, &stripe_ud,
+			    DSS_ULT_EC, 0, 0, NULL);
+	if (rc)
+		goto ev_out;
+	rc = ABT_eventual_wait(stripe_ud.asu_eventual, (void **)&status);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto ev_out;
+	}
+	if (*status != 0)
+		rc = *status;
+	else
+		rc = 0;
+
+ev_out:
+	ABT_eventual_free(&stripe_ud.asu_eventual);
+out:
+	return rc;
+
+}
+
+/* Driver function for full_stripe encode. Fetches the data and then invokes
+ * second function to encode the parity.
+ */
+static int
+agg_encode_local_parity(struct ec_agg_entry *entry)
+{
+	int rc = 0;
+
+	rc = agg_fetch_data_stripe(entry);
```

Comment:
This patch only handle all data stripes are local ?   The following patch may fetch the data cell remotely?

Reply:
Correct.  The full-featured patch fetches remote data for partial stripe update.